### PR TITLE
[WIP] Remove singletons to ensure rmmFinalize() really finalizes.

### LINKED ISSUE #144: [BUG] rmm.finalize not releasing memory (pool mode)
**Describe the bug**
Calling `rmm.finalize()` after rmm has been initialized in pool mode should/used to free up the memory pool. This no longer happens.

**Steps/Code to reproduce bug**
```
import rmm
from rmm import rmm_config as rmm_cfg

rmm_cfg.use_pool_allocator = True
rmm.initialize()
```
Pool allocated with 1/2 the cpu memory
```
 NVIDIA-SMI 410.104      Driver Version: 410.104      CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:3B:00.0 Off |                    0 |
| N/A   38C    P0    27W /  70W |   7669MiB / 15079MiB |      0%      Default 
```

```
rmm.finalize()
```

Gpu memory usage is still 1/2 gpu memory.
```
| NVIDIA-SMI 410.104      Driver Version: 410.104      CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:3B:00.0 Off |                    0 |
| N/A   43C    P0    28W /  70W |   7669MiB / 15079MiB |      0%      Default |
```

**Expected behavior**
Gpu memory should be freed 

**Environment details (please complete the following information):**
 - Environment location: Conda
 - Method of RMM install: Conda install nightly rmm=0.10.*
librmm-0.10.0a191007
rmm-0.10.0a191007
 - Please run and attach the output of the `rmm/print_env.sh` script to gather relevant environment details

**Additional context**

### PR DESCRIPTION
**Pull request by harrism**
The recent change to using memory resources for allocation/deallocation created a singleton for each type of device_memory_resource. The problem is they are created and initialized the first time `rmmInitialize` is called with corresponding options, but the singletons are never deleted, so `rmmFinalize()` didn't actually free memory pools. Instead of using singletons, this PR simply creates an instance of the appropriate MR in `rmmInitialize()` and stores it as the default resource using `set_default_resource`. In `rmmFinalize`, it resets the default resource to the cuda_memory_resource (the default, which is stateless), and deletes the previously set default_resource, ensuring any pool is deleted. 

This way a subsequent call to `rmmInitialize()` after `rmmFinalize()` acts just like the first.

Also makes all `cuda_memory_resource` instances compare equal, and all `managed_memory_resource` instances compare equal. (Per https://en.cppreference.com/w/cpp/memory/memory_resource/do_is_equal "Two memory_resources compare equal if and only if memory allocated from one memory_resource can be deallocated from the other and vice versa.")

Question: to avoid memory leaks, `rmmInitialize()` now does nothing if it is already initialized, and `rmmFinalize()` now does nothing if it is not initialized.  Should these instead throw exceptions / return errors? I was worried that doing so might conflict with the behavior of cuDF's RMM initialization options. @shwina can you comment?

Fixes #144 

Suggest ignoring whitespace changes when reviewing -- I did some cleanup and changed some indentation.

### CODE REVIEW

File: src/rmm.cpp [hunk 1]
```diff
@@ -59,39 +59,48 @@ const char * rmmGetErrorString(rmmError_t errcode) {
...
-    rmm::mr::device_memory_resource * memory_resource = rmm::mr::detail::initial_resource();
-
-    if (rmm::Manager::usePoolAllocator())
-    {
-        memory_resource = rmm::mr::detail::pool_resource(rmm::Manager::getOptions().initial_pool_size);
-
-    }else if(rmm::Manager::useManagedMemory()){
-      memory_resource = rmm::mr::detail::managed_resource();
-
-    }else if(rmm::Manager::useManagedMemory() && rmm::Manager::usePoolAllocator()){
```
Comment (harrism) [inline src/rmm.cpp]: Note that this previous logic was incorrect -- there would be no way to create a managed pool allocator.

Comment (jrhemstad) [reply src/rmm.cpp]: Yeah, I noticed that as well and fixed it in the same way here: https://github.com/rapidsai/rmm/pull/106 :)

File: include/rmm/mr/cnmem_memory_resource.hpp [hunk 2]
```diff
@@ -99,33 +102,32 @@ class cnmem_memory_resource : public device_memory_resource {
...
-   *---------------------------------------------------------------------------**/
+   *-------------------------------------------------------------------------**/
   void do_deallocate(void* p, std::size_t, cudaStream_t stream) override {
     auto status = cnmemFree(p, stream);
     if (CNMEM_STATUS_SUCCESS != status) {
-#ifndef NDEBUG
-      std::cerr << "cnmemFree failed \n";
-#endif
+      std::string msg = cnmemGetErrorString(status);
+      throw std::runtime_error{"cnmemFree failed: " + msg};
```
[the remark this answers is not in the document]
Comment (harrism) [reply include/rmm/mr/cnmem_memory_resource.hpp]: Thanks, will fix.

File: src/mr/default_memory_resource.cpp [hunk 3]
```diff
@@ -25,49 +25,20 @@
...
+}  // namespace anonymous
 
 device_memory_resource* get_default_resource() { return get_default().load(); }
 
 device_memory_resource* set_default_resource(
-    device_memory_resource* new_resource) {
-  new_resource = (new_resource == nullptr) ? detail::initial_resource() : new_resource;
+  device_memory_resource* new_resource) {
+  new_resource = (new_resource == nullptr) ?
+    new cuda_memory_resource() : new_resource;
```
[the remark this answers is not in the document]
Comment (harrism) [reply src/mr/default_memory_resource.cpp]: Yes, I was worried about this, was waiting for other eyes to confirm the problem. I like your idea.

**Comment by shwina** [comment]
@harrism It should be OK for `rmmInitialize()` and `rmmFinalize()` to exhibit the behaviour you described. We basically force this behaviour in Python-land, i.e., if the user were to set the RMM options twice in cuDF, we ignore them the second time.

**Comment by ayushdg** [comment]
I built against 5a421397a60bf31889cc8b0d27d015f16170e571 and tested it locally. I can confirm that calling `rmm.finalize` did release memory after `rmm.initialize` was invoked with the pool mode configuration. Couple of issues I noticed:

1) Initializing rmm with pool mode for the first time did not acquire memory pool. (Need to finalize first)
An example demonstrates it better:
```
import rmm
from rmm import rmm_config

rmm_config.use_pool_allocator=True

rmm.initialize() # Does nothing
```
```
import rmm
from rmm import rmm_config

rmm_config.use_pool_allocator=True

rmm.finalize()
rmm.initialize() # Now this works as expected
```

2) Using RMM in pool mode crashed groupie (might extend to other methods but I happened to see it with groupby).
```
>>> import cudf
>>> cudf.set_allocator(pool=True)
>>> df = cudf.DataFrame()
>>> df['a'] = [1,2,3]
>>> df['b'] = [4,5,6]
>>> df.groupby('a').max()
terminate called after throwing an instance of 'std::runtime_error'
  what():  cnmemFree failed: CNMEM_STATUS_INVALID_ARGUMENT
Aborted (core dumped)
```

**Comment by harrism** [comment]
> @harrism It should be OK for `rmmInitialize()` and `rmmFinalize()` to exhibit the behaviour you described. We basically force this behaviour in Python-land, i.e., if the user were to set the RMM options twice in cuDF, we ignore them the second time.

@shwina I think we need to resolve the issue that @ayushdg witnessed, namely having to call `rmm.finalize()` before calling `rmm.initialize()`. Perhaps `rmm.initialize()` should first finalize? Would that be more or less surprising than having to do it manually?

File: include/rmm/detail/memory_manager.hpp [hunk 4]
```diff
@@ -31,223 +31,218 @@
...
+    Manager& operator=(const Manager&) = delete;
+    std::mutex manager_mutex;
+    std::set<cudaStream_t> registered_streams;
+    Logger logger;
+
+    rmmOptions_t options;
+    bool is_initialized;
+
+    // unique pointers to all allocated resources
+    std::unique_ptr<rmm::mr::device_memory_resource> memory_resource{};
```
Comment (jrhemstad) [inline include/rmm/detail/memory_manager.hpp]: Consider renaming to make it more descriptive of what this is: it's a pointer to the resource that was created in initialization. 
```suggestion
    std::unique_ptr<rmm::mr::device_memory_resource> initialized_resource{};
```

File: include/rmm/detail/memory_manager.hpp [hunk 5]
```diff
@@ -31,223 +31,218 @@
...
+    Manager(const Manager&) = delete;
+    Manager& operator=(const Manager&) = delete;
+    std::mutex manager_mutex;
+    std::set<cudaStream_t> registered_streams;
+    Logger logger;
+
+    rmmOptions_t options;
+    bool is_initialized;
+
+    // unique pointers to all allocated resources
```
Comment (jrhemstad) [inline include/rmm/detail/memory_manager.hpp]:
```suggestion
    // unique pointer to the resource created in initialization
```

File: src/rmm.cpp [hunk 6]
```diff
@@ -59,39 +59,29 @@ const char * rmmGetErrorString(rmmError_t errcode) {
 // Initialize memory manager state and storage.
 rmmError_t rmmInitialize(rmmOptions_t *options)
 {
-
+  if (!rmm::Manager::getInstance().isInitialized()) {
```
Comment (jrhemstad) [inline src/rmm.cpp]: I think this logic should be pushed into the `Manager::initialize` function.

File: src/rmm.cpp [hunk 7]
```diff
@@ -59,39 +59,29 @@ const char * rmmGetErrorString(rmmError_t errcode) {
...
+  }
 
-    return RMM_SUCCESS;
+  return RMM_SUCCESS;
 }
 
 // Shutdown memory manager.
 rmmError_t rmmFinalize()
 {
+  if (rmm::Manager::getInstance().isInitialized()) {
```
Comment (jrhemstad) [inline src/rmm.cpp]: Likewise, I think this logic should go in the `Manager::finalize` function.

**Comment by shwina** [comment]
We *do* finalize RMM before initializing when the user calls `cudf.set_allocator()`.

@ayushdg does groupby also crash when you enable pool mode "manually"? i.e., with:

```python
import rmm
from rmm import rmm_config

rmm_config.use_pool_allocator=True

rmm.finalize()
rmm.initialize() # Now this works as expected
```

**Comment by jrhemstad** [comment]
We really need to get `clang-format` set up. The formatting changes are really distracting and made resolving merge conflicts in https://github.com/rapidsai/rmm/pull/106 all the more difficult.

**Comment by ayushdg** [comment]
> @shwina I think we need to resolve the issue that @ayushdg witnessed, namely having to call `rmm.finalize()` before calling `rmm.initialize()`. Perhaps `rmm.initialize()` should first finalize? Would that be more or less surprising than having to do it manually?

Just to clarify, having to call `rmm.finalize()` before `rmm.initialize()` is something that I don't need to when testing rmm built from the tip of branch-0.10, but something I need to do now when building from this branch. 

> We _do_ finalize RMM before initializing when the user calls `cudf.set_allocator()`.
> 
> @ayushdg does groupby also crash when you enable pool mode "manually"? i.e., with:
> 
> ```python
> import rmm
> from rmm import rmm_config
> 
> rmm_config.use_pool_allocator=True
> 
> rmm.finalize()
> rmm.initialize() # Now this works as expected
> ```

I built rmm with the latest updates 616629b and I no longer see the groupby crash reported before. But now I see a segfaults instead.

```
>>> import rmm
>>> a = rmm.device_array([1,2,3], dtype="int64")
>>> exit()
Segmentation fault (core dumped)
```

**Comment by shwina** [comment]
Thanks @ayushdg . Trying to reproduce locally.

**Comment by shwina** [comment]
@ayushdg I think the function you are looking for is `to_device`:

```python
In [4]: rmm.to_device(np.array([1, 2, 3]))                                                                                                                                                                      
Out[4]: <numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f231f1ee9b0>
```

Investigating the issue with groupby crash (which I can reproduce)

**Comment by ayushdg** [comment]
I can still reproduce the segfault while exiting the python interpreter:
```
>>> rmm.to_device(np.array([1, 2, 3]))
<numba.cuda.cudadrv.devicearray.DeviceNDArray at 0x7f231f1ee9b0>
>>> exit()
segmentation fault (core dumped)
```

File: src/memory_manager.cpp [hunk 8]
```diff
@@ -102,45 +102,50 @@ namespace rmm
...
    * ------------------------------------------------------------------------**/
   void Manager::finalize() {
-    std::lock_guard<std::mutex> guard(manager_mutex);
-    registered_streams.clear();
-    logger.clear();
-    memory_resource.reset();
-    getInstance().is_initialized = false;
+    // finalization before initialization is a
+    if (isInitialized()) {
+      std::lock_guard<std::mutex> guard(manager_mutex);
```
[the remark this answers is not in the document]
Comment (harrism) [reply src/memory_manager.cpp]: Fixed, thanks.

**Comment by harrism** [comment]
I added some new tests and checks for errors in rmmFree to partially fix #147 (but see comments on that issue).

**Comment by shwina** [comment]
I can confirm that this PR together with https://github.com/rapidsai/cudf/pull/3038 resolves the groupby crash we were seeing earlier, but now I'm running into a new issue.

This code exits with `rmm.rmm.RMMError: RMM_ERROR_NOT_INITIALIZED`.

```python
import cudf
a = cudf.Series([1,2,3])
```

whereas this code does not:

```python
import cudf

def f():
    a = cudf.Series([1,2,3])
```

This has to do with the fact that we use atexit` to finalize RMM. In `rmm/__init__.py`, we do:

```
atexit.register(finalize)
```

In the first case, `finalize()` is called *before* all RMM objects have been destroyed, while in the second, all RMM objects have been destroyed before `finalize()` is called.

Don't know a solution yet but researching.

**Comment by shwina** [comment]
I should mention that the issue above is orthogonal to this PR. In fact, it probably always existed but we didn't know about it since `rmmFinalize()` wasn't finalizing.

**Comment by jrhemstad** [comment]
> I should mention that the issue above is orthogonal to this PR. In fact, it probably always existed but we didn't know about it since `rmmFinalize()` wasn't finalizing.

This is occurring because prior to this PR, we erroneously didn't require that RMM still be initialized in order to call `rmm::alloc` or `rmm::free`.

**Comment by shwina** [comment]
> In Python, can you not guarantee that rmmFinalize isn't called until all objects have been destroyed (i.e., all rmm::free calls have been made).

This is exactly what needs to happen, and I'm looking into the best way to achieve this.

**Comment by shwina** [comment]
I think I have a fix, I just need to clean it up and articulate it.

**Comment by shwina** [comment]
Fixed in #149

**Comment by harrism** [comment]
@kkraus14 I can tell you want to merge it, but I want to add some further documentation to this PR before it merges.

**Comment by kkraus14** [comment]
> @kkraus14 I can tell you want to merge it, but I want to add some further documentation to this PR before it merges.

Sorry, didn't see a WIP tag and saw the review approval 😄

**Pull request merged**